Mathlib/RingTheory/PowerSeries/Ideal.lean

```diff
@@ -0,0 +1,218 @@
+/-
+Copyright (c) 2025 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca, Anthony Fernandes, Marc Robin
+-/
+
+import Mathlib.RingTheory.PowerSeries.Inverse
+import Mathlib.RingTheory.PowerSeries.Trunc
+import Mathlib.RingTheory.Finiteness.Ideal
+import Mathlib.RingTheory.Noetherian.OfPrime
+
+/-!
+# Ideals in power series.
+
+We gather miscellaneous results about prime ideals in `R⟦X⟧`. More precisely, we prove that, given
+a prime ideal `I` of `R⟦X⟧`, if the ideal generated by the constant coefficients of the `f ∈ I` is
+generated by `n` elements, then `I` is generated by at most `n + 1` elements (actually it is
+generated by `n` elements if `X ∉ I` and `n + 1` elements otherwise). This implies immediately that
+`R⟦X⟧` is noetherian if `R` is. We follow the proof given in [zbMATH05171613].
+
+
+## Main results
+
+* `PowerSeries.Ideal.eq_span_insert_X_of_Xmem_of_span_eq` : given an ideal `I` of `R⟦X⟧` such that
+  `X ∈ I`, if `I.map constantCoeff` is generated by `S : Set T`, then `I` is generated by
+  `insert X (C '' S)`.
+* `PowerSeries.Ideal.exist_eq_span_eq_ncard_of_X_not_mem` : given a prime ideal `I` of `R⟦X⟧` such
+  that `X ∉ I`, if `I.map constantCoeff` is generated by a finite set `S : Set T`, then there exists
+  `T : Set R`, of the same cardinality as `S`, that generates `I`.
+* `PowerSeries.Ideal.IsPrime.FG_iff` : a prime ideal `P` of `R⟦X⟧` is finitely generated if and only
+  if `P.map constantCoeff` is finitely generated.
+
+
+## Implementation details
+
+There is a simpler proof of `IsNoetherianRing R → IsNoetherianRing R⟦X⟧`, similar to that for
+polynomials. On the other hand our proof has the advantage of giving explicitly the number of
+generators, and it will be used to prove that, for a domain `R`, we have
+`IsPrincipalIdealRing R → UniqueFactorizationMonoid R⟦X⟧`.
+
+-/
+
+variable {R : Type*}
+
+open Ideal Set Finset
+
+namespace PowerSeries.Ideal
+
+section Semiring
+
+variable [Semiring R]
+
+lemma mem_map_constantCoeff_iff {I : Ideal R⟦X⟧} {r : R} :
+    r ∈ I.map constantCoeff ↔ ∃ f ∈ I, f.constantCoeff = r :=
+  I.mem_map_iff_of_surjective _ constantCoeff_surj
+
+lemma constantCoeff_mem_map_of_mem {I : Ideal R⟦X⟧} {f : R⟦X⟧} (hf : f ∈ I) :
+    f.constantCoeff ∈ I.map constantCoeff :=
+  mem_map_constantCoeff_iff.2 ⟨_, hf, rfl⟩
+
+end Semiring
+
+variable [CommRing R] {I : Ideal R⟦X⟧} {g : R⟦X⟧} (hg : g ∈ I)
+
+section Xmem
+
+theorem map_constantCoeff_le_self_of_Xmem (hXI : X ∈ I) :
+  (C (R := R))'' I.map constantCoeff ≤ I := by
```

Comment:
```suggestion
  I.map (C.comp constantCoeff) ≤ I := by
```
I think we prefer stating it like this.
(you can get back to the original proof via `suffices ∀ x ∈ I, C (constantCoeff x) ∈ I by simpa [Ideal.map_le_iff_le_comap]` on the first line)